src/ScriptedInputSource.cpp

```diff
@@ -127,9 +161,81 @@ PlayerInputAbs ScriptedInputSource::getNextInput()
 	if (mStartTime + mWaitTime > SDL_GetTicks() && serving)
 		return {};
 
+	if(!mMatch->getBallActive())
+	{
+		mRoundStepCounter = 0;
+		setInputDelay( 0 );
+	}
+	else
+	{
+		mRoundStepCounter += 1;
+	}
+
+	// whenever the opponent touches the ball, the bot pauses for a short, random amount of time
+	// to orient itself. This time depends on the current difficulty level, and increases as the game
+	// progresses.
+	int opp_touches = state.getHitcount(RIGHT_PLAYER);
+	if( opp_touches != mOldOppTouches)
+	{
+		mOldOppTouches = opp_touches;
+		if(opp_touches != 0)
+		{
+			int base_difficulty = std::max( 0, 2 * mDifficulty - 30 );
+			int max_difficulty = std::min(75, base_difficulty + 2 * getCurrentDifficulty());
+			std::uniform_int_distribution<int> dist{base_difficulty, max_difficulty};
+			setInputDelay( dist(mRandom) );
+		}
+	}
+
+	// important: get the actual speed, not the simulated one
+	float bv_x = mMatch->getBallVelocity().x;
+	if(bv_x != mOldBallVx) {
+		mOldBallVx = bv_x;
+		std::uniform_int_distribution<int> dist{0, 100};
+		if( dist( mRandom ) < mDifficulty ) {
+			float amount = std::min(25, getCurrentDifficulty()) / 50.f + std::max(0, mDifficulty - 5) / 25.f;
+			int err_time = 25 + mDifficulty + std::min(75, getCurrentDifficulty());
+			setBallError(err_time, amount);
+		}
+	}
+
 	PlayerInputAbs raw_input{wantleft, wantright, wantjump};
 	if(mSide == RIGHT_PLAYER) {
 		raw_input.swapSides();
 	}
 	return raw_input;
 }
+
+void ScriptedInputSource::setInputDelay(int delay)
+{
+	if(delay <= 0)
+	{
+		delay = 0;
+	} else {
+		delay = std::max(delay, mReactionTime);
+	}
+
+	mReactionTime = delay;
+}
+
+int ScriptedInputSource::getCurrentDifficulty() const
+{
+	int exchange_seconds = mRoundStepCounter / 75;
+	float difficulty_effect = std::sqrt( mDifficulty );
+	// minimum game time until the bot starts making mistakes:
+	// ~5 minutes at highest difficulty, 10 seconds for very easy.
+	int min_duration = 300 - static_cast<int>(difficulty_effect * 57);
+	int offset_seconds = std::max( 0, exchange_seconds - min_duration );
+	int diff_mod = (offset_seconds * mDifficulty) / 25;
+	return diff_mod;
```

Comment:
Is it intended that this functions always return 0 on very hard?

Reply:
yes, the idea was that it is possible to disable this simulated difficulty entirely.